### PR TITLE
feat(kuma-cp): enable MeshTraffic permission on zoneEgress

### PR DESCRIPTION
**Pull request by lukidzi**
### Checklist prior to review
- [X] [Link to relevant issue][1] as well as docs and UI issues --
- [X] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Tests (Unit test, E2E tests, manual test on universal and k8s) --
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [X] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go [hunk 1]
```diff
@@ -10,15 +12,16 @@ import (
...
 	api "github.com/kumahq/kuma/pkg/plugins/policies/meshtrafficpermission/api/v1alpha1"
 	v3 "github.com/kumahq/kuma/pkg/plugins/policies/meshtrafficpermission/xds"
 	xds_context "github.com/kumahq/kuma/pkg/xds/context"
 	"github.com/kumahq/kuma/pkg/xds/generator"
 )
 
 var (
-	_   core_plugins.PolicyPlugin = &plugin{}
-	log                           = core.Log.WithName("MeshTrafficPermission")
+	_   core_plugins.EgressPolicyPlugin = &plugin{}
```
Comment (michaelbeaumont) [inline pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go:23]: Not really part of this PR but why can't this be part of `PolicyPlugin` and just have `EgressMatchedPolicies` return an empty set in the usual case? This makes it seem like it only works on egress?

File: pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go [hunk 2]
```diff
@@ -84,3 +88,46 @@ func (p plugin) Apply(rs *core_xds.ResourceSet, ctx xds_context.Context, proxy *
...
+						if err := configurer.Configure(filterChain); err != nil {
+							return err
+						}
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
```
Comment (michaelbeaumont) [inline pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go:133]: why the trailing newline removal?

**Review by michaelbeaumont — APPROVED** [review]

File: pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go (hunk 1 above)
Comment (lahabana) [reply pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go:23]: > why can't this be part of PolicyPlugin and just have EgressMatchedPolicies return an empty set in the usual case?
Main reason was not breaking the existing API with an extra method.

File: test/e2e_env/multizone/meshtrafficpermission/meshtrafficpermission.go [hunk 3]
```diff
@@ -184,6 +221,32 @@ spec:
 		Expect(err).ToNot(HaveOccurred())
 
 		// then
-		trafficAllowed()
+		trafficAllowed("test-server.mesh")
+	})
+
+	It("should allow the traffic to the external service thorugh the egress", func() {
```
Comment (lahabana) [inline test/e2e_env/multizone/meshtrafficpermission/meshtrafficpermission.go:227]:
```suggestion
	It("should allow the traffic to the external service through the egress", func() {
```

File: pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go (hunk 1 above)
Comment (lobkovilya) [reply pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go:23]: > This makes it seem like it only works on egress?

Every EgressPolicyPlugins is a PolicyPlugin:
```go
type EgressPolicyPlugin interface {
	PolicyPlugin
	// EgressMatchedPolicies returns all the policies of the plugins' type matching the external service that
	// should be applied on the zone egress.
	EgressMatchedPolicies(*core_mesh.ExternalServiceResource, xds_context.Resources) (core_xds.TypedMatchingPolicies, error)
}
```
We don't plan to have policies that work solely on Egress. 

> why can't this be part of PolicyPlugin and just have EgressMatchedPolicies return an empty set in the usual case?

Interface shows what plugin can and can't do, you don't have to go to implementation and check if it returns empty set.

**Comment by lukidzi** [comment]
I've created a new PR to 2.3 because of other changes that were after rebase to 2.3. I've applied code review changes https://github.com/kumahq/kuma/pull/7061

File: pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go (hunk 1 above)
Comment (michaelbeaumont) [reply pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go:23]: > Main reason was not breaking the existing API with an extra method.

OK yeah that makes sense for introducing it in the first place I suppose.

> We don't plan to have policies that work solely on Egress.

Yes, I know that's my point. I'm saying the name "EgressPolicy" implies to me that it works only on egress.

> Interface shows what plugin can and can't do, you don't have to go to implementation and check if it returns empty set.

I dunno, for who exactly is this useful? And either way you actually _do_ have to go to the implementation because that's the only way to see if a plugin actually implements `EgressMatchedPolicies`. Until we rewrite in Idris and have dependent types you're going to have to check the implementation to see what the plugin can and can't do.

But obviously if people feel strongly about this it doesn't matter at all.
